File: Data Structures & Algorithms/minimum-path-sum/submission-0.java
class Solution {
    int [][] dp;
    int R,C;
    public int minPathSum(int[][] grid) {
        int r=grid.length; R=r;
        int c= grid[0].length; C=c;
        dp= new int[r][c];
        for (int[] row : dp) {
            Arrays.fill(row, -1);
        }
    
        return dfs(0,0,grid);
    }
    public int dfs(int r, int c, int[][] grid){
        if(r==R-1 && c==C-1){
            return grid[r][c];
        }
        if(r==R||c==C){
            return Integer.MAX_VALUE;
        }
        if(dp[r][c]!=-1){
            return dp[r][c];
        }
        dp[r][c]= grid[r][c]+ Math.min(dfs(r+1,c,grid),dfs(r,c+1, grid));
        return dp[r][c];
    }
}

File: Data Structures & Algorithms/triangle/submission-0.cpp
class Solution {
public:
    int minimumTotal(vector<vector<int>>& triangle) {
         vector<vector<int>> memo= triangle;
        for (auto& row : memo)
        fill(row.begin(), row.end(), INT_MAX);
        return dfs(0,0, triangle, memo);
    }
    int dfs(int r, int c, vector<vector<int>>& triangle, vector<vector<int>> &memo){
        if(r>=triangle.size()){
            return 0;
        }
        if(memo[r][c]!= INT_MAX){
            return memo[r][c];
        }
        memo[r][c]=triangle[r][c]+ min(dfs(r+1, c, triangle, memo), dfs(r+1, c+1, triangle, memo));
        return memo[r][c];
    }
};